cmd/soroban-cli/src/config/key.rs

```diff
@@ -0,0 +1,148 @@
+use std::{fmt::Display, str::FromStr};
+
+use serde::{Deserialize, Serialize};
+
+use super::secret::{self, Secret};
+use crate::xdr;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Failed to extract public key from secret")]
+    SecretPublicKey,
+    #[error(transparent)]
+    Secret(#[from] secret::Error),
+    #[error(transparent)]
+    StrKey(#[from] stellar_strkey::DecodeError),
+    #[error("failed to parse key {0}")]
+    Parse(String),
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum Key {
+    Secret(Secret),
+    PublicKey { public_key: Public },
+    MuxedAccount { muxed_account: MuxedAccount },
+}
+
+impl Key {
+    pub fn public_key(&self, hd_path: Option<usize>) -> Result<xdr::MuxedAccount, Error> {
+        let bytes = match self {
+            Key::Secret(secret) => secret.public_key(hd_path)?.0,
+            Key::PublicKey {
+                public_key: Public(key),
+            } => key.0,
+            Key::MuxedAccount {
+                muxed_account: MuxedAccount(stellar_strkey::ed25519::MuxedAccount { ed25519, id }),
+            } => {
+                return Ok(xdr::MuxedAccount::MuxedEd25519(xdr::MuxedAccountMed25519 {
+                    ed25519: xdr::Uint256(*ed25519),
+                    id: *id,
+                }))
+            }
+        };
+        Ok(xdr::MuxedAccount::Ed25519(xdr::Uint256(bytes)))
+    }
+
+    pub fn private_key(
+        &self,
+        hd_path: Option<usize>,
+    ) -> Result<stellar_strkey::ed25519::PrivateKey, Error> {
+        match self {
+            Key::Secret(secret) => Ok(secret.private_key(hd_path)?),
+            _ => Err(Error::SecretPublicKey),
```

Comment:
It appears this function returns a private key, but when no private key occurs, it returns an error saying that "Failed to extract public key from secret".

Reply:
Oops!